webui/react/src/services/stream/stream.ts

```diff
@@ -0,0 +1,151 @@
+}
+
+export class Stream {
+
+    #ws: WebSocket;
+    #retries: number = 0;
+    #numSyncs: number = 0;
+
+    #entities: Array<Record<Streamable, Entity>> = [];
+    #curEntity: Record<Streamable, Entity> | undefined;
+
+    #syncSent: string | undefined = undefined;
+    #syncComplete: string | undefined = undefined;
+    #syncStarted: string |undefined = undefined;
+
+    constructor(wsUrl: string, onUpsert: (m: Record<string, any>) => void, onDelete: (s: Streamable, a: Array<number>) => void) {
+        this.#ws = new WebSocket(wsUrl);
+        this.#ws.onopen = () => {
+            console.log('Streaming websocket opened!');
+            this.advanceSubscription();
+        };
+        this.#ws.onerror = async (err) => {
+            console.log('Streaming websocket errored: ', err);
+            console.log('# of retries: ', this.#retries);
+            const wait = await this.retry();
+            if (!wait) throw err;
+            this.#ws.dispatchEvent(new Event('open'));
+        };
+
+        this.#ws.onmessage = (event) => {
+
+            const msg = JSON.parse(event.data) as Record<string, any>;
+            if (msg['sync_id']) {
+                if (!msg['complete']) {
+                    this.#syncStarted = msg['sync_id'];
+                } else {
+                    this.#syncComplete = msg['sync_id'];
+                    this.advanceSubscription();
+                }
+            } else {
+                // Ignore all messages between when we send a new subscription and when the
+                // sync-start message for that subscription arrives.  These are the online
+                // updates for a subscription we no longer care about.
+                if (this.#syncSent !== this.#syncStarted) return;
+
+                forEach(msg, (val, k) => {
+                    if (k.includes('_deleted')) {
+                        const stream_key = trimEnd(k, '_deleted') as Streamable;
+                        const deleted_keys = decode_keys(val as string);
+                        this.#curEntity?.[stream_key].keyCache.delete_msg(deleted_keys);
+                        onDelete(stream_key, deleted_keys);
+                    } else {
+                        this.#curEntity?.[StreamEntityMap[k]].keyCache.upsert([val.id], val.seq);
+                        onUpsert(msg);
+                    }
+                });
+
+            }
+
+        };
+    }
+
+    async retry(): Promise<boolean> {
+        try {
+            const backoff = backoffs[this.#retries];
+            this.#retries += 1;
+            await sleep(backoff);
+            return true;
+        } catch {
+            return false;
+        }
+    }
+
+    #sameAsCur(entity: Record<Streamable, Entity>): boolean {
+        if (!this.#curEntity) return false;
+        let same = true;
+        forEach(entity, (val, k) => {
+            if (!this.#curEntity?.[k as Streamable].spec.equals(val.spec)) same = false;
+        });
+        return same;
```

Comment:
This feels like a case for Array.prototype.every(), for the early exit behavior:

```
return Object.entries(entity).every(
    ([k, val]) => this.#curEntity![k as Streamable].spec.equals(val.spec)
)
```

Wait, but even then... this isn't quite right, because you really need to compare two sparse dictionaries.

If your `curEntity` was:
```
{ 
  projects: pspec1
  trials: tspec1
}
```

and your `entity` was:
```
{
  projects: pspec1
}
```

then this check would say "true" because it checks that all keys in `entity` match the same keys on `curEntity`, but it doesn't check for extra keys in `curEntity`.

Reply:
I named the function poorly, but the intention was that, if I'm currently monitoring
```
{ 
  projects: pspec1
  trials: tspec1
}
```
and I subscribe to 
```
{
  projects: pspec1
}
```
then I should skip sending the subscription to the stream because what I want for `projects` has not changed. Thinking about this, I wonder if backend can do something similar, suppose we want to keep the fact that there is only one live subscription, can backend parse new subscription and only make changes if necessary?
For example, given the current subscription being
 ```
{ 
  projects: pspec1
  trials: tspec1
}
```
when receiving a new subscription 
 ```
{ 
  projects: pspec2
  trials: tspec1
}
```
can we only make changes to `projects`, and keep everything `trials` untouched?

Reply:
> can backend parse new subscription and only make changes if necessary?

No.  That was the initial design, and it was way too hard to implement.

Plus, the option to allow that design created some extremely confusing corner cases in the client, like if:

- client starts subscribed to trials 1 and 2
- client receives offline messages until seqno S1
- client resubscribes to just trial 1
- server sees update for trial 2 with seqno S2
- server sees update for trial 1 with seqno S3
- client sees update to trial 1 with seqno S3, saves S3 as max_seqno
- client resubscribes to trials 1 and 2, listing both 1 and 2 as known entities and report max_seqno as S3.  This is a problem because with 1 and 2 in known and max_seqno of S3, trial 2's update with seqno S2 will never be sent.

But I do think that having multiple independent subscriptions (each one immutable) is a simple and effective alternative for solving this problem; that's why I made a ticket for that.